README.md
# sod-di

Dependency Injection for node.js

## Install

```sh
npm install sod-di
```

## Documentation

* **[API Documentation](doc/api.md)**
* **[Error Messages](doc/error-messages.md)**
* **[Error Handling](doc/error-handling.md)**

## Example

```javascript
var di = new require('sod-di')('MyDependencies');

di.register('pi').value(Math.PI);
di.register('logger').value(console);
di.register('area').factory(function(pi) {
	return function(radius) {
		return radius * radius * pi;
	};
});
di(function(logger, area) {
	logger.log(area(2)); // stdout: 12.566370614359172
});
```

## Why

Using dependency injection instead of require() in your project hugely improves and simplifies the ability to write unit tests.


## Features

* create multiple independent dependency injectors
* public / private dependencies
* import dependency injectors to access its public dependencies
* useful error handling
* 100% unit test coverage


CHANGELOG.md
# 1.2.5 (2016-03-14)

## Changes

- [npm] update lodash to ^4.6.1

# 1.2.4 (2016-03-14)

## Changes

- [fix] stack wasn't propagated on error on newer node.js versions
- [chore] remove unused and undocumented `lazy` argument in `fileFactory(file, lazy)`

# 1.2.3 (2014-12-15)

## Changes

- fix jsdoc return value on `.register()` for code completion in WebStorm
- add new method `register().get()` for immediate factory invocation

# 1.2.2 (2014-12-05)

## Changes

- propagate original stack on require error

# 1.2.1 (2014-11-28)

## Changes

- allow numbers in injector and dependency names

# 1.2.0 (2014-11-26)

## Changes

- add api `register().fileValue()`
- add api `register().fileFactory()`
- pass errors to onError if `require()` throws while `register('key', onError).[file|fileValue|fileFactory]()`
- set `enumerable: false` on internal file property for error handling

# 1.1.3 (2014-11-02)

## Changes

- Add one of three error codes as `{Number} error.code` to each error. Constants: diFactory.Error.(COULD_NOT_REQUIRE|NOT_A_FUNCTION|DEPENDENCY_NOT_FOUND)

# 1.1.2 (2014-11-02)

## Changes

- Fix di.file() error behavior - now passes every error to {function} onError


test/di.js
var sinon = require('sinon');
require('expectations');

describe("sod-di", function() {

	var diFactory = require(process.argv.indexOf('html-cov') !== -1 ? '../di-cov.js' : '../di.js');
	var testfileFnParts = [__dirname, '../fixture/dependencyFn.js'];
	var testfileFn = require('path').join(__dirname, testfileFnParts[1]);
	var testfileValue = require('path').join(__dirname, '../fixture/dependencyValue.js');
	var testfileError = require('path').join(__dirname, '../fixture/dependencyError.js');

	it("case insensitive, di self register, prefix with di name access", function() {
		var foo = diFactory('foo');
		foo(function(di, DI, fooDi, foodi, FOODI) {
			expect(di).toBe(foo);
			expect(DI).toBe(foo);
			expect(fooDi).toBe(foo);
			expect(foodi).toBe(foo);
			expect(FOODI).toBe(foo);
		});
	});

	it("register()", function() {
		var foo = diFactory('foo');
		var service = function() {
			return 1;
		};
		expect(foo.get('service')).toBe(null);
		foo.register('service').value(service);
		expect(foo.get('service')).toBe(service);
	});

	it("register().factory(fn)", function() {
		var foo = diFactory('foo');
		var value = 123;
		var serviceCalled = 0;
		var service = function(fooValue, value) {
			serviceCalled += 1;
			return fooValue + value;
		};
		foo.register('value').value(value);
		foo.register('service').factory(service);

		// fetch service twice, only executed once, gets dependencies itself
		expect(serviceCalled).toBe(0);
		expect(foo.get('service')).toBe(value + value);
		expect(foo.get('service')).toBe(value + value);
		expect(serviceCalled).toBe(1);

		// error if no method provided
		foo.register('fail', function(error) {
			expect(error instanceof Error).toBe(true);
		}).factory(null);
	});

	it("register().get()", function() {
		var foo = diFactory('foo');
		var value = 123;
		var serviceCalled = 0;
		var service = function(fooValue, value) {
			serviceCalled += 1;
			return fooValue + value;
		};
		foo.register('value').value(value);
		var result = foo.register('service').factory(service).get();
		expect(serviceCalled).toBe(1);
		expect(result).toBe(value + value);
	});

	it("require()", function() {
		var foo = diFactory('foo');

		// return instance of error to provided errback
		foo.require('fail', function(error) {
			expect(error instanceof Error).toBe(true);
		});

		// throw if no callback provided and dependency is missing
		expect(function() {
			foo.require('fail');
		}).toThrow();

		expect(foo.require('di')).toBe(foo);
	});

	it("invoke()", function() {
		var foo = diFactory('foo');
		// check if alias exists
		expect(foo).toBe(foo.invoke);
	});

	it("newChild() inheritance", function() {
		// first
		var foo = diFactory('foo');
		// inheritance through newChild (creates new di + imports itself)
		var bar = foo.newChild('bar');
		// make bar di itself public
		bar.register('di').public();
		// inheritance through imports argument
		var baz = diFactory('baz', [foo, bar]);

		// first has not imports
		expect(foo.getImportNames()).toEqual([]);
		// second has foo
		expect(bar.getImportNames()).toEqual(['foo']);
		// third got "foo" twice but should only have it once
		expect(baz.getImportNames()).toEqual(['foo', 'bar']);

		// di should be actual di, but you can access the rest by name
		baz(function(di, bazDi, barDi) {
			expect(di).toBe(baz);
			expect(bazDi).toBe(baz);
			expect(barDi).toBe(bar);
		});

		// fooDi must be null, because it is not set to public
		expect(baz.get('fooDi')).toBe(null);

		// set to public, now it is available
		foo.register('di').public();
		expect(baz.get('fooDi')).toBe(foo);
	});

	it("import cache", function() {
		var a, b;
		var value = 1;
		var factory = sinon.stub().returns(value);
		var foo = diFactory('foo');
		var bar = foo.newChild('bar');

		// do not cache null
		foo.register('stu').value(undefined);
		expect(bar.get('cached')).toBe(null);

		// cache import result
		foo.register('cached').factory(factory).public();
		a = bar.get('cached');
		b = bar.get('cached');
		expect(a).toBe(value);
		expect(a).toBe(b);
		sinon.assert.calledOnce(factory);

		// import clears cache - but value remains correct
		bar.import(diFactory('new'));
		expect(bar.get('cached')).toBe(value);
	});

	it("allow /[a-z0-9]/i as injector and dependency name", function() {
		var di = diFactory('123f456{(]´äb');
		expect(di.diName).toBe('f456b');
		di.register('123g789{(]´äc').value('1');
		expect(di.require('g789c')).toBe('1');
		expect(di.require('f456bG789c')).toBe('1');
	});

	describe("DependencyInjectionError", function() {
		it("expose error class", function () {
			expect(diFactory.DependencyInjectionError).toBeDefined();
			expect(diFactory.Error).toBeDefined();
			expect(diFactory.Error).toBe(diFactory.DependencyInjectionError);
			expect(diFactory.Error.COULD_NOT_REQUIRE).toBe(101);
			expect(diFactory.Error.NOT_A_FUNCTION).toBe(102);
			expect(diFactory.Error.DEPENDENCY_NOT_FOUND).toBe(103);
		});

		it("should be instanceof Error", function() {
			expect(new diFactory.Error() instanceof Error).toBe(true);
		});

		it("should work in non V8 environment", function () {
			var captureStackTrace = Error.captureStackTrace;
			Error.captureStackTrace = undefined;
			var error = new diFactory.Error();
			expect(error.stack).toBeTruthy();
			Error.captureStackTrace = captureStackTrace;
		});
	});

	describe("error handling", function() {
		var foo = diFactory('foo');
		var errback;
		var brokenFn = function(missingA, di, missingB) {};
		brokenFn.__di_filename = 'my/long/filename.js';

		it("di(false)", function() {
			// errback
			foo(false, false, errback = sinon.spy());
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);
			expect(errback.getCall(0).args[0].code).toBe(diFactory.Error.NOT_A_FUNCTION);

			// without errback (throw)
			expect(function() {
				foo(false);
			}).toThrow();
		});

		it("di(function(missingDependency){})", function() {
			// errback
			foo(brokenFn, false, errback = sinon.spy());
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);
			expect(errback.getCall(0).args[0].code).toBe(diFactory.Error.DEPENDENCY_NOT_FOUND);
			// put missing dependencies into the message
			expect(errback.getCall(0).args[0].message).toMatch(/"missingA, missingB"/);
			// put method context into the message
			expect(errback.getCall(0).args[0].message).toMatch(/function *\( *missingA, *di, *missingB *\)/);
			// put fn.__di_filename into the message
			expect(errback.getCall(0).args[0].message).toContain(brokenFn.__di_filename);

			// without errback (throw)
			expect(function() {
				foo(brokenFn);
			}).toThrow();
		});

		it("di.file()", function() {
			// errback
			foo.file(testfileFn, null, errback = sinon.spy());
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);
			expect(errback.getCall(0).args[0].code).toBe(diFactory.Error.DEPENDENCY_NOT_FOUND);
			expect(errback.getCall(0).args[0].message).toContain(testfileFn);

			// without errback (throw)
			expect(function() {
				foo.file(testfileFn);
			}).toThrow();

			// missing file
			foo.file('some-non-existing-file', null, errback = sinon.spy());
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);
			expect(errback.getCall(0).args[0].code).toBe(diFactory.Error.COULD_NOT_REQUIRE);
		});

		it("di.require()", function() {
			// errback
			foo.require('unknown', errback = sinon.spy());
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);
			expect(errback.getCall(0).args[0].code).toBe(diFactory.Error.DEPENDENCY_NOT_FOUND);

			// without errback (throw)
			expect(function() {
				foo.require('unknown');
			}).toThrow();
		});

		it("di.register().factory()", function() {
			// errback
			foo.require('unknown', errback = sinon.spy());
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);
			expect(errback.getCall(0).args[0].code).toBe(diFactory.Error.DEPENDENCY_NOT_FOUND);

			// without errback (throw)
			expect(function() {
				foo.register('x').factory(false);
			}).toThrow();
		});

		it("di.callback(errback)", function() {
			// errback
			var cb = foo.callback(brokenFn, false, errback = sinon.spy());
			cb();
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);
			expect(errback.getCall(0).args[0].code).toBe(diFactory.Error.DEPENDENCY_NOT_FOUND);

			// without errback (throw)
			expect(function() {
				var cb = foo.callback(brokenFn);
				cb();
			}).toThrow();
		});

		it("di.register().file()", function() {
			foo.register('file', errback = sinon.spy()).file('$unknown');
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);

			// without errback (throw)
			expect(function() {
				foo.register('file').file('$unknown');
			}).toThrow();
		});

		it("di.register().file() - add filename to stack on error", function() {
			var onError = sinon.spy();
			foo.register('depa').file(testfileFn);
			foo.register('depb').file(testfileFnParts);
			foo.require('depa', onError);
			foo.require('depb', onError);
			// two errors - broken factory + missing dependency (times two)
			sinon.assert.callCount(onError, 4);
			// first error mentions file
			expect(onError.getCall(0).args[0].message).toContain(testfileFn);
			expect(onError.getCall(2).args[0].message).toContain(testfileFn);
		});

		it("di.register().file() - propagate stack from require error", function() {
			foo.register('file', errback = sinon.spy()).fileValue(testfileError);

			expect(errback.getCall(0).args[0].stack).toContain('thisMethodDoesNotExist is not defined');
			expect(errback.getCall(0).args[0].stack).toContain('dependencyError.js');
		});

		it("di.register().fileValue()", function() {
			foo.register('file', errback = sinon.spy()).fileValue('$unknown');
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);

			// without errback (throw)
			expect(function() {
				foo.register('file').fileValue('$unknown');
			}).toThrow();
		});

		it("di.register().fileFactory()", function() {
			foo.register('file', errback = sinon.spy()).fileFactory('$unknown');
			expect(errback.getCall(0).args[0] instanceof diFactory.Error).toBe(true);

			// without errback (throw)
			expect(function() {
				foo.register('file').fileFactory('$unknown');
			}).toThrow();
		});
	});

	it("file()", function() {
		var foo = diFactory('foo');

		expect(foo.file(testfileFn, { brokenDependency: 1 })).toBe(1);
		expect(foo.file(testfileFnParts, { brokenDependency: 2 })).toBe(2);

		// return null on missing file - and throw/onError Error
		expect(foo.file('missing-file', null, function(error) {
			expect(error instanceof diFactory.Error);
		})).toBe(null);

		// return null if fn is not a function - and throw/onError diFactory.Error
		expect(foo.file(testfileValue, null, function(error) {
			expect(error instanceof diFactory.Error);
		})).toBe(null);
	});

	it("register().file() - choose factory or value on contents", function() {
		var foo = diFactory('foo');
		foo.register('brokenDependency').value('injected value');
		foo.register('value').file(testfileValue);
		foo.register('factory').file(testfileFn);
		expect(foo.require('value')).toBe('dependency value');
		expect(foo.require('factory')).toBe('injected value');
	});

	it("register().fileValue()", function() {
		var foo = diFactory('foo');
		foo.register('file').fileValue(testfileFn);
		expect(foo.get('file')).toBe(require(testfileFn));
	});

	it("register().fileFactory()", function() {
		var foo = diFactory('foo');
		var onError = sinon.spy();
		foo.register('file').fileFactory(testfileFn);
		foo.require('file', onError);
		sinon.assert.called(onError);
		expect(onError.getCall(0).args[0].message).toContain('could not inject "brokenDependency"');
	});

	it("mapInvoke", function() {
		var foo = diFactory('foo');
		var bar = diFactory('bar');
		foo.register('value').value(1);
		bar.register('value').value(2);
		var result = diFactory.mapInvoke([foo, bar], function(value) {
			return value;
		});
		expect(result).toEqual([1, 2]);
	});

	describe("showDependencies()", function () {
		it("showDependencies((result) => void) should pass dependency tree into result", function() {
			var foo = diFactory('afoo');
			foo.register('apriv').value(1);
			foo.register('apub').value(1).public();
			var bar = diFactory('bbar');
			bar.register('bpriv').value(1);
			bar.register('bpub').value(1).public();
			var baz = diFactory('cbaz');
			baz.register('cpriv').value(1);
			baz.register('cpub').value(1).public();

			var result = '' +
				'Dependency Injector: afoo\n' +
				'\n' +
				'  afoo\n' +
				'    public\n' +
				'      apub\n' +
				'    private\n' +
				'      di\n' +
				'      apriv\n' +
				'  bbar (inherited)\n' +
				'    public\n' +
				'      bpub\n' +
				'  cbaz (inherited)\n' +
				'    public\n' +
				'      cpub\n';

			foo.import([bar, baz, /* catch circular: */ foo]);

			foo.showDependencies(function(output) {
				expect(output).toBe(result);
			});

			// cancel on circular
			foo.showDependencies(function(output) {
				expect(output).toBe(undefined);
			}, true, {afoo: true});
		});

		it("showDependencies() should use console.log if first argument is missing", function() {
			var foo = diFactory('foo');
			foo.register('apriv').value(1);
			var result = '';
			foo.showDependencies(function(output) {
				result = output;
			});

			sinon.stub(console, 'log');
			foo.showDependencies();
			try {
				sinon.assert.calledOnce(console.log);
				sinon.assert.calledWithMatch(console.log, result);
			} finally {
				console.log.restore();
			}
		});
	});

	it("custom injections", function() {
		var foo = diFactory('foo');
		foo(function(di, value, vNull, vFalse) {
			expect(di).toBe(foo);
			expect(value).toBe(1);
			expect(vFalse).toBe(false);
			// null becomes changed to undefined as null itself is already the value for missing dependency
			expect(vNull).toBe(undefined);
		}, { value: 1, vNull: null, vFalse: false });
	});
});


test/dependencyFn.js
module.exports = function(brokenDependency) { return brokenDependency };

test/brokenDependency.js
/* istanbul ignore next */
module.exports = function(brokenDependency) {}

di.js
var path = require('path');
var _ = require('lodash');
var getParameterNames = require('get-parameter-names');

var factoryKey = '__di_factory';
var fileKey = '__di_filename';
var limitCharacters = /[^a-z0-9]+/ig;
var appendedNumbers = /^[0-9]+/;

/**
 * @param {string[]} args
 * @param {object} dependencies
 * @returns {string}
 */
function getMissingDependencies(args, dependencies) {
	return _.remove(_.map(args, function(value, index) {
		return dependencies[index] === null ? value : null;
	})).join(', ');
}

/**
 * @param {function} [errorHandler]
 * @param {diFactory.Error|DiError} error
 * @throws
 */
function handleError(errorHandler, error) {
	if(typeof errorHandler !== 'function') {
		throw error;
	}
	errorHandler(error);
}

/**
 * @param {string|string[]} file
 * @param {function} onError
 * @param {string} diName
 * @returns {*}
 * @private
 */
function getFileContents(file, onError, diName) {
	var contents;
	file = _.isArray(file) ? path.join.apply(path, file) : file;
	try {
		contents = require(file);
		typeof contents === 'function' && Object.defineProperty(contents, fileKey, {
			enumerable: false,
			value: file
		});
		return contents;
	} catch(error) {
		handleError(onError, new DiError(DiError.COULD_NOT_REQUIRE, diName, 'Could not require file', null, error));
		return null;
	}
}

/**
 * @param {string} name - accessible via `.diName` and be able to request a specific
 *      service from a parent di by using this prefix like di(function(prefixService) {});
 * @param {di|di[]} [imports] - parent di to lock for, if dependency was not found in this one
 * @returns {di}
 */
function diFactory(name, imports) {
	// sanitized injector name in lowercase
	var diName = String(name).replace(limitCharacters, '').replace(appendedNumbers, '').toLowerCase();

	// public and private dependencies - key is lowercase: { name: value } and { <di-name>name: value }
	var _registers = {};

	// imported di instances
	var _imports = [];

	// once a dependency was found in an import, { name: value } is saved here ... is being cleared on .import()
	var _importCache = {};

	// public dependency names - key is lowercase: { name: true } and { <di-name>name: true }
	var _public = {};

	// public and private dependency names for .showDependencies() - key is lowercase: { name: true }
	var _names = {};

	/**
	 * dependency inject on fn
	 * @name di
	 * @throws Error if fn is not a function
	 * @throws Error if at least one dependency is missing
	 * @param {function} fn
	 * @param {object} [custom] overwrite injections with custom values
	 * @param {function(err)} [onError] - callback to catch errors, if no callback is provided,
	 *      this method throws on error on error
	 * @returns {*}
	 */
	function di(fn, custom, onError) {
		var args, injections;

		if(typeof fn !== 'function') {
			handleError(onError, new DiError(DiError.NOT_A_FUNCTION, diName, 'fn must be typeof function'));
			return null;
		}

		args = getParameterNames(fn);
		injections = custom ? args.map(function(arg) {
			// allow every custom value besides null (change null => undefined)
			return custom.hasOwnProperty(arg) ? (custom[arg] === null ? undefined : custom[arg]) : di.get(arg);
		}) : args.map(di.get);

		if(injections.indexOf(null) !== -1) {
			handleError(onError, new DiError(DiError.DEPENDENCY_NOT_FOUND, diName, 'could not inject "' + getMissingDependencies(args, injections) + '" (either missing or not setPublic() if imported)', fn));
			return null;
		}

		return fn.apply(fn, injections);
	}

	/**
	 * alias
	 * @type {di}
	 */
	di.invoke = di;

	/**
	 * alias
	 */
	di.diFactory = diFactory;

	/**
	 * dependency inject on file if module.exports is function
	 * @name di
	 * @throws Error if fn is not a function
	 * @throws Error if at least one dependency is missing
	 * @param {string|string[]} file - if file === string[], then path.join(file)
	 * @param {object} [custom] overwrite injections with custom values
	 * @param {function(err)} [onError] - callback to catch errors, if no callback is provided,
	 *      this method throws on error on error
	 * @returns {*}
	 */
	di.file = function(file, custom, onError) {
		var fn;
		if((fn = getFileContents(file, onError, diName)) === null) {
			return null;
		}
		if(typeof fn !== 'function') {
			handleError(onError, new DiError(DiError.NOT_A_FUNCTION, diName, 'require("' + file + '") did not return a function'));
			return null;
		}
		return di(fn, custom, onError);
	};

	/**
	 * @param {di|di[]} imports
	 * @returns {di}
	 */
	di.import = function(imports) {
		_.each([].concat(imports), function(d) {
			if(d && d.diName && _imports.indexOf(d) === -1 && d !== di) {
				_imports.push(d);
				_importCache = {};
			}
		});
		return di;
	};

	/**
	 * print dependencies to console.log or fn (if provided)
	 * @param {function(output)} [fn]
	 * @param {boolean} [inherited=false] - only show public
	 * @param {object} [processed] - already processed di names
	 */
	di.showDependencies = function(fn, inherited, processed) {
		var indent = function(str) { return '      ' + str; }
		var add = function(str) { str && lines.push(str); }
		var lines = [];

		processed = processed || {};
		if(processed.hasOwnProperty(diName)) {
			return;
		}
		processed[diName] = true;

		if(!inherited) {
			lines.push('Dependency Injector: ' + diName, '');
		}

		lines.push('  ' + diName + (inherited ? ' (inherited)' : ''));
		lines = lines.concat('    public', _.intersection(_.keys(_names), _.keys(_public)).map(indent));

		if(!inherited) {
			lines = lines.concat('    private', _.difference(_.keys(_names), _.keys(_public)).map(indent));
		}

		_.invokeMap(_imports, 'showDependencies', add, true, processed);

		;(fn || console.log)(lines.join('\n') + (inherited ? '' : '\n'));
	};

	/**
	 * get internal dependency / invoke factory
	 * @param {string} key in lowercase
	 * @param publicOnly
	 * @param onError
	 * @returns {*|null}
	 */
	function getDependency(key, publicOnly, onError) {
		if(_registers.hasOwnProperty(key) && (publicOnly !== true || _public[key])) {
			if(_registers[key] && _registers[key][factoryKey]) {
				_registers[key] = _registers[key]({ __di_onError: onError });
			}
			return _registers[key];
		}

		return null;
	}

	/**
	 * @throws Error if name is a factory and factory does not find a dependency
	 * @param {string} name
	 * @param {boolean} [publicOnly]
	 * @param {function(err)} [onError] - callback to catch errors, if no callback is provided,
	 *      this method throws on error on error
	 * @returns {*|null}
	 */
	di.get = function(name, publicOnly, onError) {
		var index, dependency, key = String(name).toLowerCase();

		if((dependency = getDependency(key, publicOnly, onError)) !== null) {
			return dependency;
		}

		if(_importCache.hasOwnProperty(key)) {
			return _importCache[key];
		}

		for(index = 0; index < _imports.length; index += 1) {
			if((dependency = _imports[index].get(key, true, onError)) !== null) {
				return _importCache[key] = dependency;
			}
		}

		return null;
	};

	/**
	 * same as get(), but throws an error, if name could not become resolved
	 * @throws Error if dependency was not found
	 * @param {string} name
	 * @param {function(err)} [onError] - callback to catch errors, if no callback is provided,
	 *      this method throws on error on error
 	 * @returns {*}
	 */
	di.require = function(name, onError) {
		var dependency;
		if((dependency = di.get(name, null, onError)) === null) {
			handleError(onError, new DiError(DiError.DEPENDENCY_NOT_FOUND, diName, '"' + name + '" required, but not registered'));
			return null;
		}

		return dependency;
	};

	/**
	 * @param {string} name
	 * @param {function(err)} [onError] - callback to catch errors, if no callback is provided,
	 *      this method throws on error on error
	 * @constructor
	 */
	function Dependency(name, onError) {
		this.di = di;
		this.name = String(name).replace(limitCharacters, '').replace(appendedNumbers, '').toLowerCase();
		this.onError = onError;
	}

	/**
	 * set dependency public
	 * @returns {Dependency}
	 */
	Dependency.prototype.public = function() {
		_public[this.name] = _public[diName + this.name] = true;
		return this;
	};

	/**
	 * @param {*} value
	 * @returns {Dependency}
	 */
	Dependency.prototype.value = function(value) {
		_registers[this.name] = _registers[diName + this.name] = value;
		_names[this.name] = true;
		return this;
	};

	/**
	 * register a namespace, that becomes dependency injected itself and is executed once on first request
	 * @throws Error if fn is not a function
	 * @param {function} fn
	 * @returns {Dependency}
	 */
	Dependency.prototype.factory = function(fn) {
		if(typeof fn !== 'function') {
			handleError(this.onError, new DiError(DiError.NOT_A_FUNCTION, diName, '"' + name + '" was defined as factory but is not typeof function'));
			return this;
		}

		var callback = di.callback(fn);
		callback[factoryKey] = true;
		this.value(callback);
		return this;
	};

	/**
	 * require file and this.factory() if module.exports is function, otherwise this.value()
	 * @param {string|string[]} file - if file === string[], then path.join(file)
	 * @returns {Dependency}
	 */
	Dependency.prototype.file = function(file) {
		var contents;
		if((contents = getFileContents(file, this.onError, diName)) !== null) {
			this[typeof contents === 'function' ? 'factory' : 'value'](contents);
		}
		return this;
	};

	/**
	 * require file and this.value() its contents
	 * @param {string|string[]} file - if file === string[], then path.join(file)
	 * @returns {Dependency}
	 */
	Dependency.prototype.fileValue = function(file) {
		var contents;
		if((contents = getFileContents(file, this.onError, diName)) !== null) {
			this.value(contents);
		}
		return this;
	};

	/**
	 * require file and this.factory() its contents
	 * @param {string|string[]} file - if file === string[], then path.join(file)
	 * @returns {Dependency}
	 */
	Dependency.prototype.fileFactory = function(file) {
		var contents;
		if((contents = getFileContents(file, this.onError, diName)) !== null) {
			this.factory(contents);
		}
		return this;
	};

	/**
	 * returns value of current dependency
	 * By default, factory() and fileFactory() lazy invoke on first use. You can enforce its
	 * evaluation by calling this method directly after registration:
	 *      di.register('foo').fileFactory('myFactory.js').get();
	 * @returns {*|null}
	 */
	Dependency.prototype.get = function() {
		return getDependency(this.name, false, this.onError);
	};

	/**
	 * register a namespace
	 * @param {string} name
	 * @param {function} [onError]
	 * @returns {Dependency}
	 */
	di.register = function(name, onError) {
		return new Dependency(name, onError);
	};

	/**
	 * @param {string} name
	 * @param {di|di[]} [imports]
	 * @returns {di}
	 */
	di.newChild = function(name, imports) {
		return diFactory(name, [di].concat(imports || []));
	};

	/**
	 * @returns {string[]}
	 */
	di.getImportNames = function() {
		return _.map(imports, function(d) {
			return d.diName;
		});
	};

	/**
	 * returns a callback that does not execute fn immediately
	 * @param {function} fn
	 * @param {object} [custom]
	 * @param {function(err)} [onError]
	 * @returns {Function}
	 */
	di.callback = function(fn, custom, onError) {
		/**
		 * @param {object} [opt]
		 * @param {function(err)} [opt.__di_onError] - internal pass error handler on lazy loading
		 */
		return function() {
			return di.invoke(fn, custom, (arguments[0] && arguments[0].__di_onError) || onError);
		};
	};

	di.diName = diName;
	di.import(imports);
	di.register('di').value(di);
	return di;
}

/**
 * invoke fn with each di in "dis" and return array of results
 * @param {di[]} dis
 * @param {function} fn
 * @returns {*[]}
 */
diFactory.mapInvoke = function(dis, fn) {
	return _.invokeMap(dis, 'invoke', fn);
};

/**
 * @name DependencyInjectionError
 */
var DiError = diFactory.Error = diFactory.DependencyInjectionError = (function() {
	/**
	 * @param {number} code
	 * @param {string} name
	 * @param {string} message
	 * @param {function} [context]
	 * @param {string} [context.__di_filename]
	 * @param {Error} [error] inherit stack from existing error
	 * @constructor
	 */
	function DependencyInjectionError(code, name, message, context, error) {
		this.code = code;
		this.message = this.name + ': ' + message + ' (di: ' + name + ')';
		this.message += context && context[fileKey] ? '\n    at file (' + context[fileKey] + ':1:0)' : '';
		this.message += context ? [].concat('', '... context ...', context.toString().split('\n').splice(0, 5), '...', '').join('\n') : '';
		if(error instanceof Error) {
			this.stack = this.message + '\n' + error.stack;
		} else if(Error.captureStackTrace) {
			Error.captureStackTrace(this, DependencyInjectionError);
		} else {
			this.stack = (new Error()).stack;
		}
	}

	DependencyInjectionError.prototype = Error.prototype;

	DependencyInjectionError.COULD_NOT_REQUIRE = 101;
	DependencyInjectionError.NOT_A_FUNCTION = 102;
	DependencyInjectionError.DEPENDENCY_NOT_FOUND = 103;

	return DependencyInjectionError;
}());

module.exports = diFactory;
